promql/engine.go

```diff
@@ -1106,6 +1106,10 @@ func (enh *EvalNodeHelper) resetBuilder(lbls labels.Labels) {
 // for each series, then passed to each call funcCall.
 func (ev *evaluator) rangeEval(prepSeries func(labels.Labels, *EvalSeriesHelper), funcCall func([]parser.Value, [][]EvalSeriesHelper, *EvalNodeHelper) (Vector, annotations.Annotations), exprs ...parser.Expr) (Matrix, annotations.Annotations) {
 	numSteps := int((ev.endTimestamp-ev.startTimestamp)/ev.interval) + 1
+	if numSteps > ev.maxSamples {
+		numSteps = ev.maxSamples
+	}
+
```

Comment:
How does this relate to the rest of the PR?

Reply:
Setting numSteps = maxSamples avoids allocating a slice via getFPointSlice/getHPointSlice that is larger than necessary.

I guess it is only thematically related to this PR, in that it tries to make a `rangeEval(Q, [UnixEpoch, Now])` avoid pointless work. Without this consideration of min(numSteps, ev.maxSamples), I believe we would allocate a maxSamples vector, which in the example can be very large indeed (some GiB). 

I think this could probably be a PR on its own, but it is related ot this one. Let me know if you'd like it split off.

Reply:
Where would this allocation take place?  `getFPointSlice` and `getHPointSlice` cap at 5,000.